cryptol-saw-core/src/Verifier/SAW/Cryptol.hs

```diff
@@ -1953,38 +2031,518 @@ exportRecordValue fields v =
+
+        scNthInjection :: Int -> Term -> IO Term
+        scNthInjection n x = do
+                             y <- scLeft (fst (tps!!n)) (snd (tps!!n)) x
+                             scInjectRight (n-1) y
+
+      return scNthInjection
+
+  -------------------------------------------------------------
+  -- Create the definition for each constructor:
+  defn_eachCtor <-
+    flip mapM (zip argTypes_eachCtor ctors) $
+        \(argTypes, ctor)->
+        do
+        let
+          ctorName   = C.ecName ctor
+          ctorNumber = C.ecNumber ctor
+          numArgs    = length argTypes
+
+        -- NOTE: we don't add the constructor arguments to the Env, as
+        -- the only references to these arguments are in the generated
+        -- SAWCore code.
+
+        -- create the vars (& names) for constructor arguments:
+        paramVars <- reverse <$> mapM (scLocalVar sc) [0..numArgs-1]
+        let paramNames = map (\x-> Text.pack ("x" ++ show x)) [0..numArgs-1]
+
+        -- create the constructor:
+        ctorBody <- addTypeAbstractions
+                    =<< scLambdaList sc (zip paramNames argTypes)
+                    =<< scNthInjection ctorNumber
+                    =<< scTuple sc paramVars
+        checkSAWCoreTypeChecks sc (C.nameIdent ctorName) ctorBody Nothing
+        return (ctorName, ctorBody)
+
+  return defn_eachCtor
+
+
+-- | importCase - translates a Cryptol case expr to SAWCore: an application
+--   of the generated SAWCore <NAME>__case function to appropriate arguments.
+--
+--   - Note missing functionality: (FIXME)
+--     - not implemented yet: default case alternatives that bind the whole scrutinee,
+--       i.e.,
+--
+--       >  case S of
+--       >    C1 ... -> E
+--       >    y      -> E[y]  -- y == S   <- NOT IMPLEMD YET
+--
+importCase ::
+  (HasCallStack) =>
+  SharedContext -> Env ->
+  C.Type -> C.Expr -> Map C.Ident C.CaseAlt -> Maybe C.CaseAlt -> IO Term
+importCase sc env b scrutinee altsMap mDfltAlt =
+  do
+  let scrutineeTy = fastTypeOf (envC env) scrutinee
+  (nm,ctors,tyParams,tyArgs) <- case scrutineeTy of
+      C.TNominal (NominalType {ntName=nm, ntParams=tyParams, ntDef=C.Enum ctors})
+                 tyArgs ->
+          return (nm,ctors,tyParams,tyArgs)
+      _ ->
+          panic "importCase"
+            [ "`case` expression scrutinee does not have an Enum type"
+            , pretty scrutineeTy
+            ]
+
+  -- Create a sequential set of `C.CaseAlt`s that exactly match the constructors:
+  --   - preconditions:
+  --      Assume `altsMap` is valid, thus not checking for extraneous
+  --      entries. (Call panic if a missing alternative is not covered
+  --      by presence of default in `mDfltAlt`.)
+
+  -- First, define what to do if alternative for constructor is missing:
+  let
+      -- | useDefaultAlt - when constructor (ctor) has no 'CaseAlt',
+      -- create a ctor specific alt-function from the mDfltAlt
+      -- "default expr".
```

Comment:
Can you elaborate a bit more (using pseudocode) on what it means to create a constructor-specific default expression? At first glance, I would have expected the default expression to be the same for each constructor. What parts of this are constructor-specific?